talk/morelanguage/morestl.tex

```diff
@@ -25,6 +25,121 @@ \subsection{More STL}
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[17]{Some new STL types \texttt{std::string\_view}}
+    \begin{itemize}
+    \item Non owning view of a continuous char sequence
+    \item Doesn't allocate memory, similar interface to std::string
+    \item Easy to copy, faster for some calls eg. substr (O(1) vs O(n))
+    \item Please ensure that the string pointed to outlives the string\_view!
+    \end{itemize}
+    \begin{exampleblock}{Some example uses}
+      \begin{cppcode*}{}
+        std::string_view sv {"Some example"};
+        auto first = sv.substr(0,sv.find_first_of(" "));
+        constexpr std::string_view location {"Geneva,CH"};
+        std::string some_string {"foo bar"};
+        std::string_view sv_str(some_string);
+        char foo[3] = {'f','o','o'};
+        std::string_view carr(foo, std::size(foo));
+      \end{cppcode*}
+    \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::optional}
+  \begin{itemize}
+  \item Indicates the absence of a value with \texttt{std::nullopt}
+  \item Has value semantics (Copy, Move, Compare, stack alloc.)
+  \item Useful in place of pointers where value semantics are intuitive
+  \end{itemize}
+  \begin{exampleblock}{Code example}
+    \begin{cppcode*}{}
+      std::optional<Phone> parse_phone(string_view in) {
+        if (is_valid_phone(in)) {
+          return in;// equiv. to optional<Phone>(in);
+        }
+        return {}; // default constructs std::nullopt
+      }
+      ...
+      auto v = parse_phone(get_input());
+      if (v) {  // alternatively v.is_valid()
+        process_phone(v.value()); // *v is equivalent
+      }
+    \end{cppcode*}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[17]{std::variant}
+  \begin{itemize}
+  \item Expresses intent that variable holds one of the types
+  \item Makes it easy to implement visitor pattern
+  \end{itemize}
+
+  \begin{exampleblock}{Code example}
+    \begin{cppcode*}{}
+      using option_t = std::variant<int,float,string>;
+      option_t opt; // default constructs first type
+      std::variant<std::monostate, NonDefault, int> myvar;
+      opt = 100;
+      int ival = std::get<int>(opt);//also std::get<0>(opt)
+      try {
+        float val = std::get<float>(opt) // will throw...
+      } catch (const std::bad_variant_access& ex) {...}
+
+      // Much easier with get_if; returns a ptr or null
+      if (const float* val = std::get_if<float>(&opt))
+          std::cout<<"option type is f64, value="<< val;
```

Comment:
Not sure I like to reintroduce pointers here. On top you're printing the pointer not the value.

Reply:
I kind of wanted to mention alternatives to simply try catch using the get api... but we anyway introduce the visitor pattern in the next slide, so this can be skipped